profit_analysis/prices.py

```diff
@@ -0,0 +1,182 @@
+                        10
+                        ** await self.get_decimals_from_token(self._token_base_address)
+                    )
+                    self._token_target_decimals = (
+                        10 ** await self.get_decimals_from_token(token_target_address)
+                    )
+                    token_n = await self.is_target_token0_or_token1()
+                    self._is_target_token0_or_token1 = token_n
+                return self
+            except Exception as e:
+                print(f"Error ({trials}/{n_trials}), retrying  create  -  {e}")
+                sleep(0.05)
+        W3.rotate_rpc_url()
+        return await self.create(token_target_address)
+
+    async def is_target_token0_or_token1(self):
+        if await self._pair.functions.token0().call() == self._token_target_address:
+            return 0
+        elif await self._pair.functions.token1().call() == self._token_target_address:
+            return 1
+        else:
+            raise Exception(
+                f"Target token ({self._token_target_address}) not in contract pair {self._pair}"
+            )
+
+    async def get_price_at_block(self, block_number: Union[int, float]):
+        trials = 0
+        n_trials = 3
+        while trials < n_trials:
+            trials += 1
+            try:
+                if self._token_target_address == self._token_base_address:
+                    price = 1.0
+                else:
+                    reserves = await self._pair.functions.getReserves().call(
+                        block_identifier=int(block_number)
+                    )
+                    if self._is_target_token0_or_token1 == 0:
+                        token_target_reserves = reserves[0]
+                        token_base_reserves = reserves[1]
+                    else:
+                        token_target_reserves = reserves[1]
+                        token_base_reserves = reserves[0]
+                    price = (
+                        (float(token_base_reserves) / float(token_target_reserves))
+                        * self._token_target_decimals
+                        / self._token_base_decimals
+                    )
+
+                price = float(price)
+                self.block_to_price[block_number] = price
+                return price
+            except Exception as e:
+                print(
+                    f"Error ({trials}/{n_trials}), retrying get_price_at_block  - {e}"
+                )
+                sleep(0.05)
+        W3.rotate_rpc_url()
+        await self.create(self._token_target_address)
+        return await self.get_price_at_block(block_number)
+
+
+async def safe_get_price(pricer, block, max_concurrency_semaphore):
+    async with max_concurrency_semaphore:
+        return await pricer.get_price_at_block(block)
+
+
+async def get_uniswap_historical_prices(
+    target_blocks,
+    token_address,
+    chain=POLYGON_CHAIN,
+    max_concurrency=1,
```

Comment:
What are optimal values here for `max_concurrency` and `block_batch_size`? @0xpanoramix

Reply:
Max concurrency -> 10 is enough, we'll update later on is required.
Block batch size -> 1024, we'll use the stride to properly manage the number of logs retrieved at each request.

Reply:
Thanks :white_check_mark: